### PR TITLE
(#2007) Adds Get-LatestGitHubRelease

### PR DESCRIPTION
**Pull request by JPRuskin**
## Description
Adds a script to address the updates to GitHub Release parsing. 

This has been kept simple, not addressing that wasn't needed for this issue.

## Motivation and Context
As discussed in #2007, the way we were using for scraping GitHub releases in this repository has changed / become unsupported. This function is now available in AU via Update_All (and within the repository). It allows us to simplify update scripts that were previously scraping GitHub releases.

This PR is a prerequisite for the numerous package updates for that issue.

## How Has this Been Tested?
- `Get-LatestGitHubRelease` has been tested against public releases
- `Get-LatestGitHubRelease` has been tested against private releases with a token
- `Get-LatestGitHubRelease` has been used with a PAT hoping to not get rate limited
- Tested using this in Update.ps1 scripts via Update_All.ps1
- Tested with all affected packages in issue #2007 (though with changes to those scripts)

## Types of changes

- ~[ ] Bug fix (non-breaking change which fixes an issue)~
- [x] New feature (non-breaking change which adds functionality)
- ~[ ] Breaking change (fix or feature that would cause existing functionality to change)~
- ~[ ] Migrated package (a package has been migrated from another repository)~

## Checklist:

- [x] My code follows the code style of this repository.
- ~[ ] My change requires a change to documentation (this usually means the notes in the description of a package).~
- ~[ ] I have updated the documentation accordingly (this usually means the notes in the description of a package).~
- ~[ ] I have updated the package description and it is less than 4000 characters.~
- ~[ ] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)~
- ~[ ] The added/modified package passed install/uninstall in the chocolatey test environment.~
- ~[ ] The changes only affect a single package (not including meta package).~

### CODE REVIEW

**Comment by RedBaron2** [comment]
@JPRuskin 
Two questions about your function.
1) Will this require a plain text API key be needed in the package update script?
2) The function only checks the latest tag, but what about a specific tag?

You wouldn't happen to know if this is something newly implemented by GitHub that created this issue?
Thanks for any insight.

**Comment by JPRuskin** [comment]
With regards to your first question, it shouldn't require us adding something in plaintext to each script - it follows examples in other update scripts (and the main Update_All.ps1) assuming there's an environment variable specific in our build - I should definitely double check that, though.

On the second, I made the choice to not complicate the function more than it needed to be for the issue it was addressing (I've got a horrible habit of adding everything and the kitchen sink, sometimes, and I'm trying to be better) - hence the name `Get-LatestGitHubRelease`.  
We could certainly extend it, if we found a requirement we wanted to address, or add another function for `Get-GitHubRelease` with a tag / name parameter.

I had heard that GitHub had made some specific change, but haven't found any handy announcement being made - just saw the issue, and the sum of builds using the previous method that started failing in [the update status](https://gist.github.com/choco-bot/a14b1e5bfaf70839b338eb1ab7f8226f).  
I may be incorrect, but it's probably better practice (and I think cleaner to read and maintain) to use the REST API rather than scraping the HTML - and it's certainly fixed the build issue(s) for [all the packages I've got branches for](https://github.com/JPRuskin/chocolatey-coreteampackages/branches/active).

**Comment by JPRuskin** [comment]
We could add further improvement to this by outputting an object with properties we guarantee rather than the direct return, but that feels a bit more than should be required (assuming breaking changes to the API don't happen frequently, which should at least be less frequently than changes to scraping the page)

File: scripts/Get-LatestGitHubRelease.ps1 [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      [string]$Owner,
+
+      # Repository name
+      [Parameter(Mandatory, Position=1)]
+      [string]$Name,
+
+      # GitHub token, used to reduce rate-limiting or access private repositories (needs repo scope)
+      [string]$Token = "$($env:github_api_key)"
+  )
+  end {
```
Comment (AdmiringWorm) [inline scripts/Get-LatestGitHubRelease.ps1:22]: shouldn't this be the following instead?

```suggestion
  process {
```

File: scripts/Get-LatestGitHubRelease.ps1 [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      $Request = @{
+          Uri = "https://api.github.com/repos/$Owner/$Name/releases/latest"
```
Comment (AdmiringWorm) [inline scripts/Get-LatestGitHubRelease.ps1:24]: Not so sure I like this being hard coded for the latest release only.
Any chance to make it allow a tag being specified by the user, but use `latest` by default?

File: scripts/au_extensions.psm1 [hunk 3]
```diff
@@ -5,6 +5,7 @@
 # but the file containing the functions is expected
 # to be named using the same name.
 $funcs = @(
+  'Get-LatestGitHubRelease'
```
Comment (AdmiringWorm) [inline scripts/au_extensions.psm1:8]: If the tag is made configurable, this should be changed to be named `Get-GitHubRelease` instead.

**Review by AdmiringWorm — CHANGES_REQUESTED** [review]

**Comment by AdmiringWorm** [comment]
> You wouldn't happen to know if this is something newly implemented by GitHub that created this issue?

They made changes to how the Artifacts are showing up on the releases page.
Previously it was embedded into the HTML, but was recently changed to be dynamically injected through javascript.

Since we typically use `Invoke-WebRequest -UseBasicParsing`, javascript is not run in this case and the artifacts never show up.

File: scripts/Get-LatestGitHubRelease.ps1 (hunk 2 above)
Comment (JPRuskin) [reply scripts/Get-LatestGitHubRelease.ps1:24]: I would almost suggest that we have a separate function for getting a specific release tag - though at the moment, we don't have any use for that functionality in this repository (I think?). 

Would that be suited to a follow up PR / issue, where we have a use for it?

Comment (RedBaron2) [reply scripts/Get-LatestGitHubRelease.ps1:24]: @JPRuskin
FYI: The FreeCAD package uses the tag _weekly-build_

File: scripts/au_extensions.psm1 (hunk 3 above)
Comment (RedBaron2) [reply scripts/au_extensions.psm1:8]: This so happens to be the same naming as one of the current cmdlet from the [PowerShellForGitHub](https://github.com/microsoft/PowerShellForGitHub) module

**Comment by RedBaron2** [comment]
This is true.
> Since we typically use `Invoke-WebRequest -UseBasicParsing`, javascript is not run in this case and the artifacts never show up.

Changes to the cmdlet `Invoke-WebRequest` which makes the switch `UseBasicParsing` a moot point.

> [-UseBasicParsing](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-webrequest?view=powershell-7.2#-usebasicparsing)
> This parameter has been deprecated. Beginning with PowerShell 6.0.0, all Web requests use basic parsing only. This parameter is included for backwards compatibility only and any use of it has **no effect** on the operation of the cmdlet.

**Comment by AdmiringWorm** [comment]
> Changes to the cmdlet Invoke-WebRequest which makes the switch UseBasicParsing a moot point.

This typically requires IE to have been executed at least once (maybe edge on Windows 11), as it uses the IE javascript engine. This is typically not the case with CI providers, as well as it takes longer for web requests to happen (from what I remember).
As such, it is not a solution.

Remember, it is also just deprecated for PowerShell 6.0, while we are using Windows PowerShell 5.0 in the repository and the CI.

**Comment by AdmiringWorm** [comment]
In the interest of time, I have updated this PR with the ability to get releases based on the tag name (defaulting to the latest release if no tag name is specified).

I'll get this merged once the appveyor build has completed.

**Comment by AdmiringWorm** [comment]
@JPRuskin your changes have been merged, thanks for your contribution 👍

**Comment by pauby** [comment]
Thank you @JPRuskin and @AdmiringWorm!

**Pull request merged**